sync15-adapter/src/state.rs

```diff
@@ -46,3 +73,568 @@ impl GlobalState {
+                        // If the server doesn't have a `crypto/keys`, start over
+                        // and reupload our `meta/global` and `crypto/keys`.
+                        ErrorKind::NoCryptoKeys { .. } => {
+                            let previous_global = mem::replace(&mut self.state.global, None);
+                            Ok(FreshStartRequired { previous_global })
+                        }
+                        _ => Err(err),
+                    },
+                }
+            }
+
+            Ready => Err(ErrorKind::UnexpectedSetupState.into()),
+
+            FreshStartRequired { previous_global } => {
+                // Wipe the server.
+                self.client.wipe_all()?;
+                self.state.global = None;
+                self.state.keys = None;
+                self.engine_state_changes.push(EngineStateChange::ResetAll {
+                    except: HashSet::new(),
+                });
+
+                // Upload a fresh `meta/global`...
+                let new_global = BsoRecord {
+                    id: "global".into(),
+                    collection: "meta".into(),
+                    modified: 0.0.into(), // Doesn't matter.
+                    sortindex: None,
+                    ttl: None,
+                    payload: new_global_from_previous(previous_global)?,
+                };
+                self.client.put_meta_global(&new_global)?;
+
+                // ...And a fresh `crypto/keys`. Note that we'll update the
+                // global state when we go around the state machine again,
+                // not here.
+                let new_keys = CollectionKeys::new_random()?.to_encrypted_bso(&self.root_key)?;
+                self.client.put_crypto_keys(&new_keys)?;
+
+                Ok(InitialWithLiveToken)
+            }
+        }
+    }
+
+    /// Runs through the state machine to the ready state.
+    pub fn to_ready(&mut self) -> error::Result<()> {
+        let mut state = InitialWithLiveToken;
+        loop {
+            match &state {
+                Ready => {
+                    self.sequence.push(mem::discriminant(&state));
+                    break;
+                }
+                // If we already started over once before, we're likely in a
+                // cycle, and should try again later. Like the iOS state
+                // machine, other cycles aren't a problem; we'll cycle through
+                // earlier states if we need to reupload `meta/global` or
+                // `crypto/keys`.
+                FreshStartRequired { .. } if self.sequence.contains(&mem::discriminant(&state)) => {
+                    return Err(ErrorKind::SetupStateCycleError.into());
+                }
+                _ => {
+                    self.sequence.push(mem::discriminant(&state));
+                    state = self.advance(state)?;
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+enum SetupState {
```

Comment:
Does this implement the flow in the PDF linked in the docs?  If so, hyperlinks appreciated.